### PR TITLE
A user should receive daily digest about total statistic

### PR DESCRIPTION
**Pull request by anatoliikryvishyn**
A user should receive an email in a specific time every day with statistic information about completed todos today and remaining todos.

### CODE REVIEW

**Comment by anatoliikryvishyn** [comment]
Hi, @serhiir.
We were trying to do automatic mail sending in production in few ways, but it needed purchasing some additional plugins.

1. The best way for us, as it was discussed, is to buy and add  [heroku scheduler](https://elements.heroku.com/addons/scheduler) to our application. This will allow us to manage easily automatic sending.
2. Other way, that we tried, was to use delayed job, but it needs separate worker on heroku, that is paid. We also tried to use [sucker punch gem](https://github.com/brandonhilkert/sucker_punch) that allows to run jobs in same dyno with application on heroku, but it is quite bad, because heroku will restart it once a day ([heroku documentation](https://devcenter.heroku.com/articles/dynos#automatic-dyno-restarts)), and as it is free dyno - it will turn to sleep mode after 1 hour inactivity. So it is bad for us.

So for now, you can test automatic sending on [local address](http://10.0.2.103:3000/), but please contact me before, so i could turn on automatic sending (It is turned off to not spam all users). Also, please, tell me how often you want to receive emails for testing? In [production server](https://trouble-tada-interlink.herokuapp.com/) you will need manually write command to send digest to users. 

After testing, please review.

File: app/logic/statistic/daily_digest.rb [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
+      daily_digest = DailyDigestView.new(mail_params)
+      receivers.each do |receiver|
+        DailyDigestMailer.statistic(daily_digest, receiver).deliver
+      end
+    end
+
+    private
+
+    def mail_params
+      {
```
Comment (serhiir) [inline app/logic/statistic/daily_digest.rb]: @anatoliikryvishyn That isn't a mail params. I would say that it's `statistics`. The name should not rely on the delivery mechanism. The name should rely on abstraction. Please rename.

File: app/logic/statistic/daily_digest.rb [hunk 2]
```diff
@@ -1,13 +1,23 @@
 module Statistic
   class DailyDigest
-    def send_to(current_user)
-      options = {
-          user: current_user,
+    def send
+      daily_digest = DailyDigestView.new(mail_params)
+      receivers.each do |receiver|
+        DailyDigestMailer.statistic(daily_digest, receiver).deliver
```
Comment (serhiir) [inline app/logic/statistic/daily_digest.rb]: @anatoliikryvishyn, I'd rename `daily_digest` as it is being used inside `DailyDigest` class (has the same name) but means different thing. What about `digest_view` or `daily_statistics`?

File: app/logic/statistic/daily_digest.rb [hunk 3]
```diff
@@ -1,13 +1,23 @@
...
+      {
           incompleted_count: Item.incompleted_count,
           completed_count: Item.completed_today_amount
       }
-      daily_digest = DailyDigestView.new(options)
-      DailyDigestMailer.statistic(daily_digest).deliver_now
+    end
+
+    def receivers
+      User.all.map {|user| user.email}
```
Comment (serhiir) [inline app/logic/statistic/daily_digest.rb]: @anatoliikryvishyn There is a performance issue: no need to load objects (with all attributes) into the memory. Use [pluck](http://apidock.com/rails/ActiveRecord/Calculations/pluck) instead.

File: app/logic/statistic/daily_digest.rb [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
+        DailyDigestMailer.statistic(daily_digest, receiver).deliver
+      end
+    end
+
+    private
+
+    def mail_params
+      {
           incompleted_count: Item.incompleted_count,
           completed_count: Item.completed_today_amount
```
Comment (serhiir) [inline app/logic/statistic/daily_digest.rb]: @anatoliikryvishyn Please reformat leading indents.

File: app/view_objects/daily_digest_view.rb [hunk 5]
```diff
@@ -18,6 +18,6 @@ def today
   end
 
   def receivers
-    @user.email
+    @receivers
```
Comment (serhiir) [inline app/view_objects/daily_digest_view.rb]: @anatoliikryvishyn It seems that the method is not being used (so the variable as well).

File: config/schedule.rb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+every 1.minute do
+  rake 'send_digest_email', :environment => 'development'
+end
```
Comment (serhiir) [inline config/schedule.rb]: @anatoliikryvishyn Please do not use such frequent period for sending.

File: config/schedule.rb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+every 1.minute do
```
Comment (serhiir) [inline config/schedule.rb]: @anatoliikryvishyn So what configuration would we use for real production?

File: test/controllers/items_controller_test.rb [hunk 8]
```diff
@@ -123,28 +123,4 @@ def setup
 
     assert_redirected_to :new_user_session
   end
-
-  test 'statistic should send email' do
-    assert_difference('ActionMailer::Base.deliveries.size', 1) do
-      post :statistic
```
Comment (serhiir) [inline test/controllers/items_controller_test.rb]: @anatoliikryvishyn Tests were deleted but no new tests were added. There are a lot of changes in the code. Please add tests.

**Comment by serhiir** [comment]
@anatoliikryvishyn Done.

File: config/schedule.rb (hunk 7 above)
Comment (anatoliikryvishyn) [inline config/schedule.rb]: @serhiir, as for me it will be good to send digest in second half of a day, so what about 5 pm ?

Comment (serhiir) [inline config/schedule.rb]: @anatoliikryvishyn I belive that depends on business domain. I'm okay with 5 pm.

**Comment by anatoliikryvishyn** [comment]
@serhiir, uodated. Please review.

**Pull request merged**